### PR TITLE
p2p, rpc: enable GetAddr, GetAddresses, and getnodeaddresses by network

### PR DESCRIPTION
**Pull request by jonatack**
This patch allows passing a network argument to CAddrMan::GetAddr(), CConnman::GetAddresses(), and rpc getnodeaddresses to return only addresses of that network.

It also contains a performance optimisation by promag to no longer call gettime for each vRandom entry inside the CAddrMan::GetAddr_() loop.

### CODE REVIEW

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -875,8 +879,13 @@ static RPCHelpMan getnodeaddresses()
     const int count{request.params[0].isNull() ? 1 : request.params[0].get_int()};
     if (count < 0) throw JSONRPCError(RPC_INVALID_PARAMETER, "Address count out of range");
 
+    const Network network{request.params[1].isNull() ? Network::NET_MAX : ParseNetwork(request.params[1].get_str())};
+    if (network == NET_UNROUTABLE) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid network (not recognized)");
```
Comment (vasild) [inline src/rpc/net.cpp:884]: Maybe print the unrecognized string:
```suggestion
        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Network not recognized: %s", request.params[1].get_str());
```

File: src/addrman.h [hunk 2]
```diff
@@ -716,13 +716,13 @@ friend class CAddrManTest;
     }
 
     //! Return a bunch of addresses, selected at random.
-    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct)
+    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct, Network net = Network::NET_MAX)
```
Comment (vasild) [inline src/addrman.h:719]: To designate "not specified / any network", maybe `std::optional` would fit better than using the special value `NET_MAX`? I plan to remove `NET_MAX` some day.

File: src/addrman.h [hunk 3]
```diff
@@ -279,7 +279,7 @@ friend class CAddrManTest;
 #endif
 
     //! Select several addresses at once.
-    void GetAddr_(std::vector<CAddress> &vAddr, size_t max_addresses, size_t max_pct) EXCLUSIVE_LOCKS_REQUIRED(cs);
+    void GetAddr_(std::vector<CAddress> &vAddr, size_t max_addresses, size_t max_pct, Network net = Network::NET_MAX) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Comment (vasild) [inline src/addrman.h:282]: The default value can be removed since all callers of this method always provide the argument.

```suggestion
    void GetAddr_(std::vector<CAddress> &vAddr, size_t max_addresses, size_t max_pct, Network net) EXCLUSIVE_LOCKS_REQUIRED(cs);
```

**Review by vasild — COMMENTED** [review]
Approach ACK.

Filtering this "server side" would help reduce network traffic especially if all addresses are requested (matters if the RPC client is not running on the same machine as `bitcoind`). It does not add any (measurable) computational overhead on the "server side".

File: src/addrman.h (hunk 2 above)
Comment (jonatack) [reply src/addrman.h:719]: I was asking myself if `std::optional` would be better here. Thanks!

File: src/rpc/net.cpp (hunk 1 above)
Comment (jonatack) [reply src/rpc/net.cpp]: Good idea! Done.

File: src/addrman.h (hunk 3 above)
Comment (jonatack) [reply src/addrman.h]: Indeed! Done.

Comment (jonatack) [reply src/addrman.h]: Indeed! Done.

**Comment by jonatack** [comment]
Thanks for the early feedback @vasild! Updated and ready for review.

File: src/addrman.h [hunk 4]
```diff
@@ -715,14 +723,20 @@ friend class CAddrManTest;
...
-    //! Return a bunch of addresses, selected at random.
-    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct)
+    /**
+     * Return all or many randomly selected addresses, optionally by network.
+     *
+     * @param[in] max_addresses  Maximum number of addresses to return (0 = all).
+     * @param[in] max_pct        Maximum percentage of addresses to return (0 = all).
+     * @param[in] network        Select only addresses of this network (nullopt = all).
+     */
+    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct, std::optional<Network> network = std::nullopt)
```
Comment (vasild) [inline src/addrman.h:733]: It just occurred to me that the only (non test) caller of `GetAddr()` is `GetAddresses()` which will always pass 3 arguments. Maybe consider removing the default from here since it is not necessary.

```suggestion
    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct, std::optional<Network> network)
```

File: test/functional/rpc_net.py [hunk 5]
```diff
@@ -187,43 +187,54 @@ def test_service_flags(self):
...
-        assert_raises_rpc_error(-8, "Address count out of range", self.nodes[0].getnodeaddresses, -1)
+        # Test for the absence of onion and I2P addresses.
+        for network in ["onion", "i2p"]:
+            assert_equal(self.nodes[0].getnodeaddresses(0, network), [])
 
-        # addrman's size cannot be known reliably after insertion, as hash collisions may occur
-        # so only test that requesting a large number of addresses returns less than that
-        LARGE_REQUEST_COUNT = 10000
-        node_addresses = self.nodes[0].getnodeaddresses(LARGE_REQUEST_COUNT)
-        assert_greater_than(LARGE_REQUEST_COUNT, len(node_addresses))
```
Comment (vasild) [inline test/functional/rpc_net.py:226]: Why remove this test?

File: doc/release-notes.md [hunk 6]
```diff
@@ -130,9 +136,6 @@ Changes to Wallet or GUI related settings can be found in the GUI or Wallet sect
 
 - Passing an invalid `-rpcauth` argument now cause bitcoind to fail to start.  (#20461)
 
-- The `getnodeaddresses` RPC now returns a "network" field indicating the
-  network type (ipv4, ipv6, onion, or i2p) for each address.  (#21594)
```
Comment (vasild) [inline doc/release-notes.md:134]: This move is unrelated to this PR. If this PR does not fly, when the move will not fly either.

File: test/functional/rpc_net.py (hunk 5 above)
Comment (jonatack) [reply test/functional/rpc_net.py:226]: It's redundant with the assertion in line 215 (before this change they were two identical tests as well, by two different authors).

File: doc/release-notes.md (hunk 6 above)
Comment (jonatack) [reply doc/release-notes.md:134]: I think it's ok. If this PR doesn't make the release, I can move the existing note to the right place when the notes are moved to the wiki for editing (I usually make some corrections in the wiki before the notes are moved back to the repo for -final).

**Review by vasild — APPROVED** [review]
ACK ac01ec096bc5ff437db5c4797e3d98b7a6efad23

(would be happy to re-ACK if the default value of the 3rd arg of `GetAddr()` is removed)

File: src/addrman.h (hunk 4 above)
Comment (jonatack) [reply src/addrman.h:733]: Good idea, doing (this will add fuzz coverage for change too).

**Comment by jonatack** [comment]
Thank you @vasild. Removed the defaults and improved the fuzzing per `git diff ac01ec0 46933d8`.

**Comment by jonatack** [comment]
CI failure is unrelated (https://github.com/bitcoin/bitcoin/issues/21632).

**Comment by jonatack** [comment]
Rebased for CI fix on master.

**Comment by vasild** [comment]
The latest push removed the default value for the `network` argument from `CAddrMan::GetAddr()` and from `CConnMan::GetAddresses()` too. The latter deserves a default value because now there are a lot of call sites like `connman.GetAddresses(..., std::nullopt)`, no?

File: src/addrman.cpp [hunk 7]
```diff
@@ -501,8 +503,10 @@ void CAddrMan::GetAddr_(std::vector<CAddress>& vAddr, size_t max_addresses, size
         assert(mapInfo.count(vRandom[n]) == 1);
 
         const CAddrInfo& ai = mapInfo[vRandom[n]];
-        if (!ai.IsTerrible())
+        // Filter for quality, and if a network is passed, filter also by network.
+        if (!ai.IsTerrible() && (network == std::nullopt || ai.GetNetClass() == network)) {
```
Comment (promag) [inline src/addrman.cpp:507]: 8d4b14e001ec026f346c6b4b00509adae7769415

I'd use a different approach to not mix different conditions:

```cpp
// Filter by network (optional)
if (network != std::nullopt && ai.GetNetClass() != network) continue;

// Filter for quality
if (ai.IsTerrible()) continue;

vAddr.push_back(ai);
```

Also note that it first filters by network, which I think is slightly better.

I also want to point that `IsTerrible` default argument is computed in each iteration, I'd suggest:

```cpp
    const int64_t now = GetAdjustedTime();

    // gather a list of random nodes, skipping those of low quality
    for (unsigned int n = 0; n < vRandom.size(); n++) {
       ...
       if (ai.IsTerrible(now)) continue;

```

File: src/addrman.h [hunk 8]
```diff
@@ -723,14 +723,20 @@ friend class CAddrManTest;
...
-    //! Return a bunch of addresses, selected at random.
-    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct)
+    /**
+     * Return all or many randomly selected addresses, optionally by network.
+     *
+     * @param[in] max_addresses  Maximum number of addresses to return (0 = all).
+     * @param[in] max_pct        Maximum percentage of addresses to return (0 = all).
+     * @param[in] network        Select only addresses of this network (nullopt = all).
+     */
+    std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct, std::optional<Network> network)
```
Comment (promag) [inline src/addrman.h:733]: 8942494e8c5eff8a3e268fe830868faf8ce19f9f

Why not `network = std::nullopt`? It would avoid changes in call sites.

**Review by promag — COMMENTED** [review]
Concept ACK.

I think this change could be simplified by setting the default argument of the network filter.

File: src/addrman.h (hunk 8 above)
Comment (jonatack) [reply src/addrman.h:733]: Yes, that is what the PR initially did, as it was a smaller diff and I tend to start with the smaller change. It is now passed explicitly in reponse to review feedback. I agree that explicit is good too.

Comment (promag) [reply src/addrman.h:733]: You can dismiss my comment, just checked @vasild comment https://github.com/bitcoin/bitcoin/pull/21843#discussion_r625554482.

**Review by vasild — APPROVED** [review]
ACK bdc57fc120affcd255b5b79b505433f5ed4dc843

Whether to have a default value for the argument of `CAddrMan::GetAddr()` and `CConnMan::GetAddresses()` is not a showstopper for this PR, IMO.

Now there are two non-test callers of `GetAddresses()` that explicitly pass `nullopt`: 

https://github.com/bitcoin/bitcoin/blob/bdc57fc120affcd255b5b79b505433f5ed4dc843/src/net.cpp#L2691

https://github.com/bitcoin/bitcoin/blob/bdc57fc120affcd255b5b79b505433f5ed4dc843/src/net_processing.cpp#L3580

File: src/addrman.cpp (hunk 7 above)
Comment (jonatack) [reply src/addrman.cpp:507]: Thanks @promag, done and added a commit by you for the gettime optimization.

**Comment by jonatack** [comment]
Thanks @vasild, agreed. Making them explicit was more work, but I reckon sooner or later someone is going to make them explicit so may as well do it now.

**Comment by jonatack** [comment]
Only change is @promag's suggestion: ` git diff bdc57fc 9d17f30`
```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index 473ac21ad0..ceab1689d7 100644
--- a/src/addrman.cpp
+++ b/src/addrman.cpp
@@ -494,6 +494,7 @@ void CAddrMan::GetAddr_(std::vector<CAddress>& vAddr, size_t max_addresses, size
+    const int64_t now{GetAdjustedTime()};
     for (unsigned int n = 0; n < vRandom.size(); n++) {
         if (vAddr.size() >= nNodes)
             break;
@@ -503,10 +504,14 @@ void CAddrMan::GetAddr_(std::vector<CAddress>& vAddr, size_t max_addresses, size
         assert(mapInfo.count(vRandom[n]) == 1);
 
         const CAddrInfo& ai = mapInfo[vRandom[n]];
-        // Filter for quality, and if a network is passed, filter also by network.
-        if (!ai.IsTerrible() && (network == std::nullopt || ai.GetNetClass() == network)) {
-            vAddr.push_back(ai);
-        }
+
+        // Filter by network (optional)
+        if (network != std::nullopt && ai.GetNetClass() != network) continue;
+
+        // Filter for quality
+        if (ai.IsTerrible(now)) continue;
+
+        vAddr.push_back(ai);
     }
 }
```

**Review by promag — COMMENTED** [review]
Code review ACK 9d17f30c7e281d0f424eb45129beb6a6e62e37b9.

**Review by vasild — APPROVED** [review]
ACK 9d17f30c7e281d0f424eb45129beb6a6e62e37b9

**Comment by jonatack** [comment]
rebased after merge of #21506, no other change

`git range-diff 4da26fb 9d17f30 ce6bca8`
```diff
     @@ src/net_processing.cpp: void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
              pfrom.vAddrToSend.clear();
              std::vector<CAddress> vAddr;
-         if (pfrom.HasPermission(PF_ADDR)) {
+         if (pfrom.HasPermission(NetPermissionFlags::Addr)) {
```

**Review by vasild — APPROVED** [review]
ACK ce6bca88e8c685c69686e0b8dc095ffc3e2ac34d

**Comment by laanwj** [comment]
Code review and lightly tested ACK ce6bca88e8c685c69686e0b8dc095ffc3e2ac34d

**Comment by jarolrod** [comment]
post-merge ACK 👍

File: src/addrman.cpp [hunk 9]
```diff
@@ -501,8 +504,14 @@ void CAddrMan::GetAddr_(std::vector<CAddress>& vAddr, size_t max_addresses, size
         assert(mapInfo.count(vRandom[n]) == 1);
 
         const CAddrInfo& ai = mapInfo[vRandom[n]];
-        if (!ai.IsTerrible())
-            vAddr.push_back(ai);
+
+        // Filter by network (optional)
+        if (network != std::nullopt && ai.GetNetClass() != network) continue;
```
Comment (sipa) [inline src/addrman.cpp:509]: Won't this result in infnite loops, if there are not enough of the requested net class?

I think doing this correctly requires keeping track of the counts per network.

Comment (jonatack) [reply src/addrman.cpp:509]: If the early exit isn't triggered (e.g. `getnodeaddresses 10000 i2p`) the loop terminates after it has iterated through vRandom.

Comment (vasild) [reply src/addrman.cpp:509]: Today's computers are so fast that even the infinite loop finishes in about 6 seconds.

Comment (jonatack) [reply src/addrman.cpp:509]: (I think the added functional tests here cover it.)
